File: api/src/vertx_clj/handlers.clj
(ns vertx-clj.handlers
  (:require [clojure.java.io :as io]
            [jsonista.core :as json]
            [vertx-clj.helpers :as h])
  (:import (io.vertx.core Handler)
           (io.vertx.ext.web RoutingContext)
           (io.vertx.ext.web.api RequestParameters)))

(def api-spec-handler
  (reify Handler
    (handle [_ routing-ctx]
      (-> ^RoutingContext routing-ctx
          .response
          (.putHeader "content-type" "application/yaml")
          (.end (-> "api.yaml"
                    io/resource
                    slurp))))))

(def health-check-handler
  (h/->json-handler
    (constantly "Yes we can! \uD83D\uDD28 \uD83D\uDD28")))

(def pipeline-create-handler
  (h/->json-handler
    #(let [params   (.get ^RoutingContext % "parsedParameters")
           group    (h/get-path-param params "group")
           name     (h/get-path-param params "name")
           pipeline (-> ^RequestParameters params
                        .body
                        .toString
                        (json/read-value h/mapper))]
       (println group name pipeline))))

(def pipeline-artifact-handler
  (reify Handler
    (handle [_ routing-ctx]
      (-> ^RoutingContext routing-ctx
          .response
          (.putHeader "content-type" "application/tar")
          (.sendFile "test.tar.gz")))))

File: apiserver/src/main/kotlin/Handlers.kt
import io.vertx.core.Future
import io.vertx.core.json.JsonObject
import io.vertx.ext.web.RoutingContext

fun toJsonResponse(routingContext: RoutingContext, content: Any): Future<Void> =
    routingContext.response()
        .putHeader("content-type", "application/json")
        .end(JsonObject(mapOf("message" to content)).encode())

fun healthCheckHandler(routingContext: RoutingContext) =
    toJsonResponse(routingContext, "Yes we can! \uD83D\uDD28 \uD83D\uDD28")

fun pipelineCreateHandler(routingContext: RoutingContext): Future<Void> {
    val params = routingContext.request().params()
    val group = params["group"]
    val name = params["name"]
    val pipeline = routingContext.bodyAsJson

    println(group)
    println(name)
    println(pipeline)

    return toJsonResponse(routingContext, pipeline)
}

fun apiSpecHandler(routingContext: RoutingContext): Future<Void> =
    routingContext.response()
        .putHeader("content-type", "application/yaml")
        .end({}.javaClass.getResource("bob/api.yaml").readText())

fun pipelineArtifactHandler(routingContext: RoutingContext): Future<Void> =
    routingContext.response()
        .sendFile("test.tar.gz")